src/java.base/share/classes/java/lang/runtime/Carrier.java

```diff
@@ -0,0 +1,1167 @@
+                        components[longIndex++], LONG_TO_DOUBLE);
+            } else if (ptype == float.class) {
+                component = MethodHandles.filterReturnValue(
+                        components[intIndex++], INT_TO_FLOAT);
+            } else if (ptype == long.class) {
+                component = components[longIndex++];
+            } else {
+                component = components[intIndex++];
+            }
+
+            MethodType methodType = methodType(ptype, Object.class);
+            reorder[index++] =
+                    MethodHandles.explicitCastArguments(component, methodType);
+        }
+
+        return List.of(reorder);
+    }
+
+    /**
+     * Given components array and index, recast and reorder that component to
+     * match shape.
+     *
+     * @param carrierShape  carrier reshape
+     * @param components    carrier components
+     * @param i             index of component to reshape
+     *
+     * @return component reshaped
+     */
+    private static MethodHandle reshapeComponent(CarrierShape carrierShape,
+                                                 MethodHandle[] components, int i) {
+        Class<?>[] ptypes = carrierShape.ptypes();
+        CarrierCounts componentCounts = CarrierCounts.tally(ptypes, i);
+        Class<?> ptype = ptypes[i];
+        int index;
+        MethodHandle filter = null;
+
+        if (!ptype.isPrimitive()) {
+            index = carrierShape.objectOffset() + componentCounts.objectCount();
+        } else if (ptype == double.class) {
+            index = carrierShape.longOffset() + componentCounts.longCount();
+            filter = LONG_TO_DOUBLE;
+        } else if (ptype == float.class) {
+            index = carrierShape.intOffset() + componentCounts.intCount();
+            filter = INT_TO_FLOAT;
+        } else if (ptype == long.class) {
+            index = carrierShape.longOffset() + componentCounts.longCount();
+        } else {
+            index = carrierShape.intOffset() + componentCounts.intCount();
+        }
+
+        MethodHandle component = components[index];
+
+        if (filter != null) {
+            component = MethodHandles.filterReturnValue(component, filter);
+        }
+
+        component = MethodHandles.explicitCastArguments(component,
+                methodType(ptype, Object.class));
+
+        return component;
+    }
+
+    /**
+     * Factory for carriers that are backed by int[] and Object[]. This strategy is
+     * used when the number of components exceeds {@link Carrier#MAX_OBJECT_COMPONENTS}.
+     */
+    private static class CarrierArrayFactory {
+        /**
+         * Lookup used to define and reference the carrier array methods.
+         */
+        private static final MethodHandles.Lookup LOOKUP;
+
+        /**
+         * Unsafe access.
+         */
+        private static final Unsafe UNSAFE;
+
+        static {
+            LOOKUP = MethodHandles.lookup();
+            UNSAFE = Unsafe.getUnsafe();
```

Comment:
It might be better to use `java.lang.invoke.MethodHandleStatics.UNSAFE`, and probably also `java.lang.invoke.MethodHandles.Lookup.IMPL_LOOKUP`.

Reply:
The package is java.lang.runtime not java.lang.invoke so both fields are not accessible.

Reply:
Well, you could use `SharedSecrets.getJavaLangInvokeAccess().findStatic(…)` and `SharedSecrets.getJavaLangInvokeAccess().findVirtual(…)` in place of `LOOKUP.findStatic(…)` and `LOOKUP.findVirtual(…)`.

Reply:
Trusted access is not needed, but I can get rid of the LOOKUP static final. The UNSAFE static final is trickier since it might impact the performance of the array getInt.